### PR TITLE
Timeout functions based on Arm CNTP

### PR DESCRIPTION
**Pull request by etienne-lms**
This P-R proposes timeout functions and an example of their use on the stm32_uart `putc()` and `flush()` function used for the platform console.

### CODE REVIEW

File: core/drivers/stm32_uart.c [hunk 1]
```diff
@@ -45,17 +49,26 @@ static vaddr_t loc_chip_to_base(struct serial_chip *chip)
...
 
 static void loc_putc(struct serial_chip *chip, int ch)
 {
 	vaddr_t base = loc_chip_to_base(chip);
+	uint64_t timeout_ref = utimeout_init(PUTC_TIMEOUT_US);
 
-	while (!(read32(base + UART_REG_ISR) & USART_ISR_TXE_TXFNF))
-		;
+	while (!(read32(base + UART_REG_ISR) & USART_ISR_TXE_TXFNF)) {
+		if (utimeout_elapsed(PUTC_TIMEOUT_US, timeout_ref)) {
```
[the remark this answers is not in the document]
Comment (jforissier) [reply core/drivers/stm32_uart.c]: > This can be tricky: the peripheral clock could be OFF, the external bus pads could be mux-ed out.

OK. That argument is convincing enough :)

> Moreover the console bus may be controlled by the non-secure world in which case secure world cannot safely enable it [...]

Sure. I was not suggesting secure world should force-enable the device.

File: core/arch/arm/kernel/delay.c [hunk 2]
```diff
@@ -44,3 +48,37 @@ void mdelay(uint32_t ms)
...
+	uint64_t origin = reference - US2CNT(us);
+	uint64_t now = read_cntpct();
+
+	if (origin < reference)
+		return now < origin || now > reference;
+
+	return now < origin && now > reference;
+}
+
+unsigned int utimeout_elapsed_us(uint32_t us, uint64_t reference)
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/delay.c]: I think the return type should be `uint32_t` instead as that's what we use as input

Comment (etienne-lms) [reply core/arch/arm/kernel/delay.c]: You're tight. I think using an `unsigned int` as input is even better.

Comment (jenswi-linaro) [reply core/arch/arm/kernel/delay.c]: That depends on which range of values we're aiming for. An advantage with `uint32_t` is that the range is well defined, while `unsigned int` is just some moderately large unsigned integer.

Comment (jforissier) [reply core/arch/arm/kernel/delay.c]: `udelay() and `mdelay()` use `uint32_t` so this type is better for consistency at least

Comment (etienne-lms) [reply core/arch/arm/kernel/delay.c]: ok. agreed.

File: core/arch/arm/kernel/delay.c [hunk 3]
```diff
@@ -44,3 +48,37 @@ void mdelay(uint32_t ms)
 {
 	udelay(1000 * ms);
 }
+
+uint64_t utimeout_init(uint32_t us)
```
Comment (jforissier) [inline core/arch/arm/kernel/delay.c]: s/utimeout_init/timeout_init_us()/ perhaps? We are not obliged to mimic `udelay()`.

File: core/arch/arm/kernel/delay.c [hunk 4]
```diff
@@ -44,3 +48,37 @@ void mdelay(uint32_t ms)
...
+{
+	return read_cntpct() + US2CNT(us);
+}
+
+bool utimeout_elapsed(uint32_t us, uint64_t reference)
```
Comment (jforissier) [inline core/arch/arm/kernel/delay.c]: Why give `us` again? Why not just `return read_cntpct() > reference;`? The system counter is guaranteed to start from zero and not to roll-over before 40 years (Arm ARM G5.1.2) so there is no need to consider the wrap-around case.
Then there is no reason to keep the `u` prefix if there is nothing relating to microseconds.
My proposal:
```C
boot timeout_elapsed(uint64_t exp)
{
	return read_cntpct() > exp;
}
```

And since the functions are very simple, they could go in `kernel/delay.h` as `static inline` IMO.

File: core/arch/arm/kernel/delay.c [hunk 5]
```diff
@@ -44,3 +48,37 @@ void mdelay(uint32_t ms)
...
+	uint64_t origin = reference - US2CNT(us);
+	uint64_t now = read_cntpct();
+
+	if (origin < reference)
+		return now < origin || now > reference;
+
+	return now < origin && now > reference;
+}
+
+uint32_t utimeout_elapsed_us(uint32_t us, uint64_t reference)
```
Comment (jforissier) [inline core/arch/arm/kernel/delay.c]: Could be simplified too. Perhaps it would be better not to introduce this function now since it is not used for the moment.

File: core/arch/arm/kernel/delay.c (hunk 3 above)
Comment (etienne-lms) [reply core/arch/arm/kernel/delay.c]: Ok, `timeout_init_us()` is fine to me.

File: core/arch/arm/kernel/delay.c (hunk 4 above)
Comment (etienne-lms) [reply core/arch/arm/kernel/delay.c]: looks ok to me. I'll fix.

File: core/arch/arm/kernel/delay.c (hunk 5 above)
Comment (etienne-lms) [reply core/arch/arm/kernel/delay.c]: indeed, this is useful for debugging only :)

File: core/arch/arm/kernel/delay.c [hunk 6]
```diff
@@ -29,12 +30,14 @@
 #include <arm.h>
 #include <kernel/delay.h>
 
+#define US2CNT(m)	(((uint64_t)m * (uint64_t)read_cntfrq()) / 1000000ULL)
+
 void udelay(uint32_t us)
 {
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/delay.c]: Since you're making changes to this function, why not simplify it by using the `timeout_init_us()` and `timeout_elapsed()` functions?

Comment (etienne-lms) [reply core/arch/arm/kernel/delay.c]: ok

File: core/drivers/stm32_uart.c [hunk 7]
```diff
@@ -45,17 +49,26 @@ static vaddr_t loc_chip_to_base(struct serial_chip *chip)
...
 static void loc_putc(struct serial_chip *chip, int ch)
 {
 	vaddr_t base = loc_chip_to_base(chip);
+	uint64_t timeout = timeout_init_us(PUTC_TIMEOUT_US);
 
-	while (!(read32(base + UART_REG_ISR) & USART_ISR_TXE_TXFNF))
-		;
+	while (!(read32(base + UART_REG_ISR) & USART_ISR_TXE_TXFNF)) {
+		if (timeout_elapsed(timeout)) {
+			if (read32(base + UART_REG_ISR) & USART_ISR_TXE_TXFNF)
```
Comment (jenswi-linaro) [inline core/drivers/stm32_uart.c]: Why is this needed?

File: core/drivers/stm32_uart.c [hunk 8]
```diff
@@ -45,17 +49,26 @@ static vaddr_t loc_chip_to_base(struct serial_chip *chip)
...
+				break;
+
```
Comment (jforissier) [inline core/drivers/stm32_uart.c]: Useless blank line

File: core/arch/arm/include/kernel/delay.h [hunk 9]
```diff
@@ -29,7 +30,22 @@
 #ifndef __KERNEL_DELAY_H
 #define __KERNEL_DELAY_H
 
+#include <stdbool.h>
+#include <stdint.h>
+
+#define US2CNT(m)	(((uint64_t)m * (uint64_t)read_cntfrq()) / 1000000ULL)
```
Comment (jforissier) [inline core/arch/arm/include/kernel/delay.h]: Could be a static inline function instead

File: core/drivers/stm32_uart.c (hunk 7 above)
Comment (etienne-lms) [reply core/drivers/stm32_uart.c]: Well... there is not really needed I guess. It allowed to still push a charater to the trace if timeout elapased but UART responded.
Ok to remove this test: would bring more questions than it could resolve issues.

File: core/drivers/stm32_uart.c (hunk 8 above)
Comment (etienne-lms) [reply core/drivers/stm32_uart.c]: ok

File: core/arch/arm/include/kernel/delay.h (hunk 9 above)
Comment (etienne-lms) [reply core/arch/arm/include/kernel/delay.h]:
```
static inline uint64_t us2cnt(uint32_t us)
{
	return ((uint64_t)us * (uint64_t)read_cntfrq()) / 1000000ULL;
}
```
Look strange a global function `us2cnt()`. Maybe `arm_cnt_us2cnt()`?

File: core/arch/arm/include/kernel/delay.h [hunk 10]
```diff
@@ -33,14 +33,17 @@
...
 #include <stdint.h>
 
-#define US2CNT(m)	(((uint64_t)m * (uint64_t)read_cntfrq()) / 1000000ULL)
-
 void udelay(uint32_t us);
 void mdelay(uint32_t ms);
 
+static inline uint64_t arm_cnt_us2cnt(uint32_t us)
+{
+	return ((uint64_t)us * (uint64_t)read_cntfrq()) / 1000000ULL;
```
Comment (jenswi-linaro) [inline core/arch/arm/include/kernel/delay.h]: Here's a dependency on `<arm.h>`, please include `<arm.h>` above.

**Comment by jenswi-linaro** [comment]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by jforissier** [comment]
@etienne-lms you haven't changed the names of the functions in the commit message... ;-o

**Comment by etienne-lms** [comment]
ok, sorry, I will fix, thanks.

**Comment by jforissier** [comment]
IBART failure is unrelated and Travis has succeeded (although GitHub reports the build is in progress).
Merging this.
